sql/core/src/main/scala/org/apache/spark/sql/execution/Exchange.scala

```diff
@@ -488,6 +488,12 @@ private[sql] case class EnsureRequirements(sqlContext: SQLContext) extends Rule[
   }
 
   def apply(plan: SparkPlan): SparkPlan = plan.transformUp {
+    case operator @ Exchange(partitioning, child, _) =>
+      child.children match {
+        case Exchange(childPartitioning, baseChild, _)::Nil =>
```

Comment:
The key difference about putting it in the planner instead of in the optimizer is that you can actually plan the child first and check its actual output ordering instead of doing this specific check for `Aggregate`.